src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c

```diff
@@ -385,6 +625,165 @@ static unsigned short clock_get_freq_impl(struct clock *clock, int idx)
+		 * requested frequency then nothing to do.
+		 */
+		if (!force && (find_matching_freq_config(curr_freq[i],
+		    acap_frequency_table, ARRAY_SIZE(acap_frequency_table)) == idx)) {
+			CLOCK_INFO(clock, "current freq and new freq are the "
+			    "same, skip updating.");
+			continue;
+		}
+
+		err = clock_wiz_busy(clock, i, 20, 50);
+		if (err)
+			break;
+		/*
+		 * Simplified formula for ACAP clock wizard.
+		 * 1) Set DIVCLK_EDGE, DIVCLK_LT and DIVCLK_HT to 0;
+		 * 2) Set CLKFBOUT_FRACT_EN to 0;
+		 * 3) O = floor(4320/freq_req), M = floor((O*freq_req)/33.333);
+		 * 4) CLKFBOUT_EDGE = if M%2 write 0x17, else write 0x16
+		 * 5) CLKFBOUT_LT_HT = (M-M%2)/2_(M-M%2)/2
+		 * 6) check CLKOUT0_PREDIV2, CLKOUT0_P5EN == 0
+		 * 7) CLKOUT0_EDGE O%2 write 0x13, else write 0x12
+		 * 8) CLKOUT0_LT_HT = (O-(O%2))/2
+		 */
+		/* Step 1) */
+		val = reg_rd(clock->clock_bases[i] + OCL_CLKWIZ_DIVCLK);
+		divclk = (struct acap_divclk *)&val;
+		divclk->divclk_edge = 0;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_DIVCLK, val);
+
+		val = reg_rd(clock->clock_bases[i] + OCL_CLKWIZ_DIVCLK_TS);
+		divclk_ts = (struct acap_divclk_ts *)&val;
+		divclk_ts->divclk_lt = 0;
+		divclk_ts->divclk_ht = 0;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_DIVCLK_TS, val);
+
+		/* Step 2) */
+		val = reg_rd(clock->clock_bases[i] + OCL_CLKWIZ_CLKFBOUT_FRACT);
+		fract = (struct acap_clkfbout_fract *)&val;
+		fract->clkfbout_fract_en = 0;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKFBOUT_FRACT, val);
+
+		/* Step 3) */
+		M = acap_frequency_table[idx].config0;
+		O = acap_frequency_table[idx].config2;
+
+		/* Step 4) */
+		val = reg_rd(clock->clock_bases[i] + OCL_CLKWIZ_CLKFBOUT);
+		clkfbout = (struct acap_clkfbout *)&val;
+		clkfbout->clkfbout_edge = (M % 2) ? 1 : 0;
+		clkfbout->clkfbout_en = 1;
+		clkfbout->clkfbout_mx = 1;
+		clkfbout->clkfbout_prediv2 = 1;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKFBOUT, val);
+
+		/* Step 5) */
+		val = 0;
+		clkfbout_ts = (struct acap_clkfbout_ts *)&val;
+		clkfbout_ts->clkfbout_lt = (M - (M % 2)) / 2;
+		clkfbout_ts->clkfbout_ht = (M - (M % 2)) / 2;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKFBOUT_TS, val);
+
+		/* Step 6, 7) */
+		val = reg_rd(clock->clock_bases[i] + OCL_CLKWIZ_CLKOUT0);
+		clkout0 = (struct acap_clkout0 *)&val;
+		clkout0->clkout0_edge = (O % 2) ? 1 : 0;
+		clkout0->clkout0_mx = 1;
+		clkout0->clkout0_used = 1;
+		clkout0->clkout0_prediv2 = 0;
+		clkout0->clkout0_p5en = 0;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKOUT0, val);
+
+		/* Step 8) */
+		val = 0;
+		clkout0_ts = (struct acap_clkout0_ts *)&val;
+		clkout0_ts->clkout0_lt = (O - (O % 2)) / 2;
+		clkout0_ts->clkout0_ht = (O - (O % 2)) / 2;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKOUT0_TS, val);
+
+		/* init the freq change */
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_INIT_CONFIG, 0x3);
+		err = clock_wiz_busy(clock, i, 100, 100);
+		if (err) {
+			/* ??? should we restore back if fail */
```

Comment:
Yeah, seems we already did in line 859? Or at least we should log a better error message?

Reply:
There is no equivalent rollback procedure for ACAP Versal yet, the best way here might be just bail out with waning message from clock_wiz_busy.

Reply:
OK, fair enough